subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/PluginDeclaration.java

```diff
@@ -134,4 +136,13 @@ public boolean equals(Object obj) {
     public int hashCode() {
         return Objects.hash(name, id, implementationClass);
     }
+
+    private void writeObject(java.io.ObjectOutputStream out) throws IOException {
+        DeprecationLogger.deprecateIndirectUsage("Declaring PluginDeclaration as an input or serializing PluginDeclaration")
```

Comment:
Following from https://github.com/gradle/gradle/pull/20050#discussion_r822420738

> There is currently a problem that the deprecation warning also is emitted when using configuration caching, so I am not sure I can include the deprecation.

Hmm, there's no way to opt out of serialization of an object?
Detecting who is serializing (e.g. from stack trace) might be an overkill, right?
Or adding a flag around (isConfigurationCacheSerialization) that is accessible statically and set/reset during serialization.

Reply:
I could leave some ugly special cases in some places. Do you have a particular use-case in mind for serializing `PluginDeclaration`? Or is this more a general comment that we could actually emit a deprecation warning before we remove the `Serializable` interface in Gradle 8?

Reply:
Just in general. Your warnings have been really useful in advancing the ecosystem.